examples/zephyr/certificate_provisioning/src/main.c

```diff
@@ -141,12 +141,12 @@ int main(void) {
         client = golioth_client_create(&client_config);
 
         golioth_client_register_event_callback(client, on_client_event, NULL);
-
-        k_sem_take(&connected, K_FOREVER);
     } else {
         LOG_ERR("Error reading certificate credentials from filesystem");
     }
 
+    k_sem_take(&connected, K_FOREVER);
```

Comment:
This will deadlock if the cert and key don't exist in the filesystem, because we never start the Golioth client thread (line 141) that eventually calls the callback that posts to this semaphore.

Reply:
I think that's the desired behavior. The shell still continues to function and you can load the device certs and reboot the board. What's blocked by this semaphore is the Hello messages that should only be sent once an connection is established.

Reply:
The behavior is desired, but I think we can do it in a cleaner and more explicit way. It looks strange to wait on a semaphore that can never be be posted to, and it's non-obvious what the desired behavior is from reading the code. Especially in example code, where we'll have lots of people who are unfamiliar with our SDK reading our code, I think we can benefit from being more explicit and less clever. 

I think having a call to `k_sleep()` in an infinite loop in the `else` case, or setting a flag that skips the log in the existing infinite loop would be clearer to users reading the code for the first time.

Reply:
I guess the difference here is we know that we'll fail to connect because we check for the files first. But I see your point. Let's a file ticket and merge this PR 👍

Reply:
https://github.com/golioth/firmware-issue-tracker/issues/395